### PR TITLE
Wait for insertion in distributed

### PR DESCRIPTION
**Pull request by KochetovNicolai**

### CODE REVIEW

**Comment by ekonkov** [comment]
autotest

File: dbms/src/Storages/Distributed/DistributedBlockOutputStream.cpp [hunk 1]
```diff
@@ -123,6 +156,40 @@ void DistributedBlockOutputStream::writeToLocal(const Block & block, const size_
...
+    for (const auto & dir_name : dir_names)
+    {
+        auto & monitor = storage.requireDirectoryMonitor(dir_name);
+        auto & pool = monitor.getPool();
+        auto connection = pool->get();
+
+        CurrentMetrics::Increment metric_increment{CurrentMetrics::DistributedSend};
+
+        if (timeout_exceeded)
+            throw Exception("Timeout exceeded. Inserted blocks: " + std::to_string(blocks_inserted), ErrorCodes::TIMEOUT_EXCEEDED);
```
Comment (alexey-milovidov) [inline dbms/src/Storages/Distributed/DistributedBlockOutputStream.cpp]: Probably there will be many similar messages for each shard. But these messages doesn't contain shard name and indistinguishable.

Comment (alexey-milovidov) [inline dbms/src/Storages/Distributed/DistributedBlockOutputStream.cpp]: What about to add another ProfileEvent for this?

File: dbms/src/Storages/Distributed/DistributedBlockOutputStream.cpp [hunk 2]
```diff
@@ -123,6 +156,40 @@ void DistributedBlockOutputStream::writeToLocal(const Block & block, const size_
...
+            remote.write(block);
+            remote.writeSuffix();
+        }
+        catch (Exception & exception)
+        {
+            std::string message = "\nWhile insertion to ";
+            message += connection->getDescription();
+            message += " Inserted blocks: " + std::to_string(blocks_inserted);
+            exception.addMessage(message);
+            LOG_ERROR(log, message);
```
Comment (alexey-milovidov) [inline dbms/src/Storages/Distributed/DistributedBlockOutputStream.cpp]: If we rethrow, don't need to log the message.

File: dbms/src/Storages/Distributed/DistributedBlockOutputStream.h [hunk 3]
```diff
@@ -21,10 +30,12 @@ class StorageDistributed;
 class DistributedBlockOutputStream : public IBlockOutputStream
 {
 public:
-    DistributedBlockOutputStream(StorageDistributed & storage, const ASTPtr & query_ast, const ClusterPtr & cluster_);
+    DistributedBlockOutputStream(StorageDistributed & storage, const ASTPtr & query_ast, const ClusterPtr & cluster_, bool insert_sync_, UInt64 insert_timeout_ = 0);
 
     void write(const Block & block) override;
 
+    void writePrefix() override { deadline = std::chrono::system_clock::now() + std::chrono::seconds(insert_timeout); }
```
Comment (alexey-milovidov) [inline dbms/src/Storages/Distributed/DistributedBlockOutputStream.h]: Deadline is initialized twice. Do we have any places where `writePrefix` is not called?
Maybe we could remove useless and misleading initialization from constructor.

File: dbms/src/Storages/Distributed/DistributedBlockOutputStream.cpp [hunk 4]
```diff
@@ -123,6 +156,40 @@ void DistributedBlockOutputStream::writeToLocal(const Block & block, const size_
 }
 
 
+void DistributedBlockOutputStream::writeToShardDirect(const Block & block, const std::vector<std::string> & dir_names, std::atomic<bool> & timeout_exceeded)
```
Comment (alexey-milovidov) [inline dbms/src/Storages/Distributed/DistributedBlockOutputStream.cpp]: Why `timeout_exceeded` is not const reference?

File: dbms/src/Storages/Distributed/DistributedBlockOutputStream.h [hunk 5]
```diff
@@ -36,10 +47,18 @@ class DistributedBlockOutputStream : public IBlockOutputStream
 
     void writeToShard(const Block & block, const std::vector<std::string> & dir_names);
 
+    void writeToShardDirect(const Block & block, const std::vector<std::string> & dir_names, std::atomic<bool> & timeout_exceeded);
```
Comment (alexey-milovidov) [inline dbms/src/Storages/Distributed/DistributedBlockOutputStream.h]: Missing comment.

File: dbms/src/Storages/Distributed/DistributedBlockOutputStream.h [hunk 6]
```diff
@@ -36,10 +47,18 @@ class DistributedBlockOutputStream : public IBlockOutputStream
...
+
 private:
     StorageDistributed & storage;
     ASTPtr query_ast;
     ClusterPtr cluster;
+    bool insert_sync = true;
+    UInt64 insert_timeout = 1;
```
Comment (alexey-milovidov) [inline dbms/src/Storages/Distributed/DistributedBlockOutputStream.h]: This default value probably have no meaning.

File: dbms/src/Storages/StorageDistributed.cpp [hunk 7]
```diff
@@ -245,7 +245,7 @@ BlockOutputStreamPtr StorageDistributed::write(const ASTPtr & query, const Setti
 
     /// DistributedBlockOutputStream will not own cluster, but will own ConnectionPools of the cluster
     return std::make_shared<DistributedBlockOutputStream>(
-        *this, rewriteInsertQuery(query, remote_database, remote_table), cluster);
+        *this, rewriteInsertQuery(query, remote_database, remote_table), cluster, settings.insert_distributed_sync, settings.insert_distributed_timeout);
```
Comment (alexey-milovidov) [inline dbms/src/Storages/StorageDistributed.cpp]: It means, that default value of last argument of constructor of DistributedBlockOutputStream is not really needed.

File: dbms/tests/integration/test_sync_insert_into_distributed/test.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+        node1.query('''
+        SET insert_distributed_sync = 1, insert_distributed_timeout = 0;
+        INSERT INTO distributed_table SELECT today() as date, number as val FROM system.numbers''', timeout=1)
+
+
+if __name__ == '__main__':
+    with contextmanager(started_cluster)() as cluster:
+        for name, instance in cluster.instances.items():
+            print name, instance.ip_address
+        raw_input("Cluster created, press any key to destroy...")
```
Comment (alexey-milovidov) [inline dbms/tests/integration/test_sync_insert_into_distributed/test.py]: Thank you for test!

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
As discussed.

File: dbms/src/Storages/Distributed/DistributedBlockOutputStream.h [hunk 9]
```diff
@@ -21,25 +32,56 @@ class StorageDistributed;
...
+
+    void calculateRemoteJobsCount();
+
+    ThreadPool::Job createWritingJob(std::vector<bool> & done_jobs, std::atomic<unsigned> & finished_jobs_count,
+                                     std::condition_variable & cond_var, const Block & block, size_t job_id,
+                                     const Cluster::ShardInfo & shard_info, size_t replica_id);
+
+    void writeToLocal(const Blocks & blocks, size_t & finished_writings_count);
+
+    /// Returns the number of blocks was read for each cluster node. Uses during exception handling.
```
Comment (alexey-milovidov) [inline dbms/src/Storages/Distributed/DistributedBlockOutputStream.h]: Written?

File: dbms/src/Storages/Distributed/DistributedBlockOutputStream.cpp [hunk 10]
```diff
@@ -1,43 +1,244 @@
...
+                                            blocks[shard_id], job_id, shards_info[shard_id], replica_id));
+            ++job_id;
+        }
+    }
+
+    const size_t jobs_count = job_id;
+    size_t finished_local_nodes_count;
+    const auto time_point = deadline;
+    auto timeout = insert_timeout;
+    auto & pool = this->pool;
```
Comment (alexey-milovidov) [inline dbms/src/Storages/Distributed/DistributedBlockOutputStream.cpp]: Looks non obvious. It's possible to capture `this`.
See also "capture with an initializer":
http://en.cppreference.com/w/cpp/language/lambda#Lambda_capture

File: dbms/src/Storages/Distributed/DistributedBlockOutputStream.cpp [hunk 11]
```diff
@@ -1,43 +1,244 @@
...
+    auto wait = [& mutex, & cond_var, time_point, & finished_jobs_count, jobs_count, & pool, timeout]()
+    {
+        std::unique_lock<std::mutex> lock(mutex);
+        auto cond = [& finished_jobs_count, jobs_count] { return finished_jobs_count == jobs_count; };
+        if (timeout)
+        {
+            if (!cond_var.wait_until(lock, time_point, cond))
```
Comment (alexey-milovidov) [inline dbms/src/Storages/Distributed/DistributedBlockOutputStream.cpp]: See "spurious wakeups" in documentation:
http://en.cppreference.com/w/cpp/thread/condition_variable/wait_until

File: dbms/src/Storages/Distributed/DistributedBlockOutputStream.cpp [hunk 12]
```diff
@@ -1,43 +1,244 @@
...
+        writeToLocal(blocks, finished_local_nodes_count);
+    }
+    catch (...)
+    {
+        exception = std::current_exception();
+    }
+
+    try
+    {
+        wait();
```
Comment (alexey-milovidov) [inline dbms/src/Storages/Distributed/DistributedBlockOutputStream.cpp]: Why we need "wait" as lambda?

File: dbms/src/Storages/Distributed/DistributedBlockOutputStream.cpp [hunk 13]
```diff
@@ -1,43 +1,244 @@
...
+        if (exception)
+            std::rethrow_exception(exception);
```
Comment (alexey-milovidov) [inline dbms/src/Storages/Distributed/DistributedBlockOutputStream.cpp]: Relation with code below is ok but very non-obvious.

File: dbms/src/Storages/Distributed/DistributedBlockOutputStream.cpp [hunk 14]
```diff
@@ -91,38 +292,71 @@ void DistributedBlockOutputStream::writeSplit(const Block & block)
...
+
+
+void DistributedBlockOutputStream::writeSplit(const Block & block)
+{
+    Blocks splitted_blocks = splitBlock(block);
+    const size_t num_shards = splitted_blocks.size();
+
     for (size_t shard_idx = 0; shard_idx < num_shards; ++shard_idx)
         if (splitted_blocks[shard_idx].rows())
             writeImpl(splitted_blocks[shard_idx], shard_idx);
```
Comment (alexey-milovidov) [inline dbms/src/Storages/Distributed/DistributedBlockOutputStream.cpp]: Should rename this method to mension async nature.

File: dbms/src/Storages/Distributed/DistributedBlockOutputStream.cpp [hunk 15]
```diff
@@ -91,38 +292,71 @@ void DistributedBlockOutputStream::writeSplit(const Block & block)
...
+    }
 
     block_io.out->writeSuffix();
 }
 
 
+void DistributedBlockOutputStream::writeToShardSync(
+    const Block & block, const Cluster::ShardInfo & shard_info, size_t replica_id)
+{
+    const auto & dir_name = shard_info.dir_names[replica_id];
```
Comment (alexey-milovidov) [inline dbms/src/Storages/Distributed/DistributedBlockOutputStream.cpp]: Not very good, that we refer to "directory names" when there are no directories.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]

**Pull request merged**